### PR TITLE
Add helper commands for debugging maps

### PR DESCRIPTION
**Pull request by JoelTroch**
This pull request is an implementation of helper commands for debugging maps as described in issue #288.

This PR is a draft because it's incomplete and I would like to have feedback on the existing work before doing the rest.

### CODE REVIEW

File: src/game/server/client.cpp [hunk 1]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+		{
+			int edicts = 0;
+			int count = 0;
+			std::vector<const char*> entities;
+
+			for (auto entity : UTIL_FindEntities()) {
+				if (!entity)
+					continue;
+
+				if (entity->edict())
```
Comment (SamVanheer) [inline src/game/server/client.cpp:681]: Any entity that doesn't have an edict is bugged so this isn't really useful. I'd expect the game to crash before this condition becomes false.

File: src/game/server/entities/player.h [hunk 2]
```diff
@@ -449,6 +455,9 @@ class CBasePlayer : public CBaseMonster
 	//For saving and level changes.
 	int m_HudColor = RGB_HUD_COLOR.ToInteger();
 
+	bool m_bInfiniteAir;
```
Comment (SamVanheer) [inline src/game/server/entities/player.h:458]: These should be save/restored to match the behavior of `FL_GODMODE` and other cheat flags.

**Review by SamVanheer — CHANGES_REQUESTED** [review]

File: src/game/server/client.cpp [hunk 3]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+
+	g_ClientCommands.Create("ent_find_by_classname", [](CBasePlayer* player, const CommandArgs& args)
+		{
+			if (args.Count() > 1)
+			{
+				int count = 0;
+				ALERT(at_console, "entindex - targetname - origin");
+
+				for (auto entity : UTIL_FindEntities()) {
+					if (!entity || !FStrEq(args.Argument(1), entity->GetClassname()))
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/client.cpp:609]: Fixed

File: src/game/server/client.cpp [hunk 4]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+
+	g_ClientCommands.Create("ent_find_by_targetname", [](CBasePlayer* player, const CommandArgs& args)
+		{
+			if (args.Count() > 1)
+			{
+				int count = 0;
+				ALERT(at_console, "entindex - classname - origin");
+
+				for (auto entity : UTIL_FindEntities()) {
+					if (!entity || !FStrEq(args.Argument(1), entity->GetClassname()))
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/client.cpp:632]: Fixed

File: src/game/server/client.cpp [hunk 5]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+				CLIENT_PRINTF(player->edict(), print_console, "usage: ent_find_by_classname <classname>\n");
+			} },
+		{.Flags = ClientCommandFlag::Cheat});
+
+	g_ClientCommands.Create("ent_find_by_targetname", [](CBasePlayer* player, const CommandArgs& args)
+		{
+			if (args.Count() > 1)
+			{
+				int count = 0;
+				ALERT(at_console, "entindex - classname - origin");
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/client.cpp:629]: Changed to print both.

File: src/game/server/client.cpp [hunk 6]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+
+				entities.push_back(entity->GetClassname());
+			}
+
+			std::sort(entities.begin(), entities.end(), [](const char* a, const char* b) {
+				return strcmp(a, b) < 0;
+				});
+
+			const char* lastClass = "";
+			for (std::vector<const char*>::iterator it = entities.begin(); it != entities.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/client.cpp:692]: Noted for future reference, replaced `ent_list` by showing all entities classname, targetname and origin with a count.

File: src/game/server/client.cpp [hunk 7]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+					count++;
+					continue;
+				}
+
+				ALERT(at_console, "Class: %s (%d)\n", lastClass, count);
+				lastClass = *it;
+				count = 1;
+			}
+
+			if (lastClass[0] != 0 && count)
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/client.cpp:703]: Noted for future reference, replaced `ent_list` by showing all entities classname, targetname and origin with a count.

File: src/game/server/client.cpp [hunk 8]
```diff
@@ -588,6 +589,124 @@ void SV_CreateClientCommands()
...
+					edicts++;
+
+				entities.push_back(entity->GetClassname());
+			}
+
+			std::sort(entities.begin(), entities.end(), [](const char* a, const char* b) {
+				return strcmp(a, b) < 0;
+				});
+
+			const char* lastClass = "";
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/client.cpp:691]: Reworked the command to print entindex, classname, targetname and origin. Pagination to be added in the future.

File: src/game/server/client.cpp (hunk 1 above)
Comment (JoelTroch) [reply src/game/server/client.cpp:681]: Removed everything "edicts count" related.

File: src/game/server/entities/player.cpp [hunk 9]
```diff
@@ -415,7 +415,8 @@ bool CBasePlayer::TakeDamage(entvars_t* pevInflictor, entvars_t* pevAttacker, fl
 	m_lastDamageAmount = flDamage;
 
 	// Armor.
-	if (0 != pev->armorvalue && (bitsDamageType & (DMG_FALL | DMG_DROWN)) == 0) // armor doesn't protect against fall or drown damage!
+	// TODO - Should m_bInfiniteArmor apply the health reduction?
```
[the remark this answers is not in the document]
Comment (JoelTroch) [reply src/game/server/entities/player.cpp:418]: Done.

File: src/game/server/entities/player.h (hunk 2 above)
Comment (JoelTroch) [reply src/game/server/entities/player.h:458]: Added both variables to the save/restore table.

File: src/game/server/entities/player.cpp [hunk 10]
```diff
@@ -429,10 +432,11 @@ bool CBasePlayer::TakeDamage(entvars_t* pevInflictor, entvars_t* pevAttacker, fl
 			flArmor = pev->armorvalue;
 			flArmor *= (1 / flBonus);
 			flNew = flDamage - flArmor;
-			pev->armorvalue = 0;
+			if (!m_bInfiniteArmor)
+				pev->armorvalue = 0;
 		}
 		else
-			pev->armorvalue -= flArmor;
+			pev->armorvalue -= m_bInfiniteArmor ? pev->armorvalue : flArmor;
```
Comment (SamVanheer) [inline src/game/server/entities/player.cpp:439]: The player will lose all armor when they take damage. This should be:
```cpp
else if (!m_bInfiniteArmor)
{
    pev->armorvalue -= flArmor;
}
```

**Review by SamVanheer — CHANGES_REQUESTED** [review]

File: src/game/server/entities/player.cpp (hunk 10 above)
Comment (JoelTroch) [reply src/game/server/entities/player.cpp]: Oops, fixed.

**Pull request merged**